core/trino-main/src/main/java/io/trino/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -64,672 +83,1650 @@
-
-                    // remote source requires nodePartitionMap
-                    NodePartitionMap nodePartitionMap = partitioningCache.apply(plan.getFragment().getPartitioning());
-                    if (groupedExecutionForStage) {
-                        checkState(connectorPartitionHandles.size() == nodePartitionMap.getBucketToPartition().length);
-                    }
-                    stageNodeList = nodePartitionMap.getPartitionToNode();
-                    bucketNodeMap = nodePartitionMap.asBucketNodeMap();
-                    bucketToPartition = Optional.of(nodePartitionMap.getBucketToPartition());
+                else if (state == DistributedStagesSchedulerState.CANCELED) {
+                    // output stage was canceled
+                    queryStateMachine.transitionToCanceled();
                 }
-
-                stageSchedulers.put(stageId, new FixedSourcePartitionedScheduler(
-                        stage,
-                        splitSources,
-                        plan.getFragment().getStageExecutionDescriptor(),
-                        schedulingOrder,
-                        stageNodeList,
-                        bucketNodeMap,
-                        splitBatchSize,
-                        getConcurrentLifespansPerNode(session),
-                        nodeScheduler.createNodeSelector(session, catalogName),
-                        connectorPartitionHandles,
-                        dynamicFilterService,
-                        tableExecuteContextManager));
-            }
-            else {
-                // all sources are remote
-                NodePartitionMap nodePartitionMap = partitioningCache.apply(plan.getFragment().getPartitioning());
-                List<InternalNode> partitionToNode = nodePartitionMap.getPartitionToNode();
-                // todo this should asynchronously wait a standard timeout period before failing
-                checkCondition(!partitionToNode.isEmpty(), NO_NODES_AVAILABLE, "No worker nodes available");
-                stageSchedulers.put(stageId, new FixedCountScheduler(stage, partitionToNode));
-                bucketToPartition = Optional.of(nodePartitionMap.getBucketToPartition());
             }
-            childStages = createChildStages.apply(bucketToPartition);
-        }
 
-        stage.addStateChangeListener(newState -> {
-            if (newState == FLUSHING || newState.isDone()) {
-                childStages.forEach(SqlStageExecution::cancel);
+            if (state == DistributedStagesSchedulerState.FAILED) {
+                StageFailureInfo stageFailureInfo = distributedStagesScheduler.getFailureCause()
+                        .orElseGet(() -> new StageFailureInfo(toFailure(new VerifyException("distributedStagesScheduler failed but failure cause is not present")), Optional.empty()));
+                ErrorCode errorCode = stageFailureInfo.getFailureInfo().getErrorCode();
+                if (shouldRetry(errorCode)) {
```

Comment:
Can we also include information in QueryStats that would allow us to easily see the number of query retry attempts, the error that was encountered before retry and time at which new attempt was scheduled ?

Reply:
There's a separate work item to update `QueryStats` (and other data structure) to reflect retries.